onnxruntime/contrib_ops/cpu/tokenizer.cc

```diff
@@ -0,0 +1,455 @@
+  size_t output_index = 0;
+  for (size_t n = 0; n < N; ++n) {
+    for (size_t c = 0; c < C; ++c) {
+      const auto* s = input_data + (n * C) + c;
+      if (mark_) {
+        new (output_data + output_index) std::string(&start_text, 1);
+        ++output_index;
+      }
+      size_t tokens = 0;
+      const size_t str_len = s->size();
+      for (size_t token_idx = 0; token_idx < str_len;) {
+        size_t tlen = utf8_bytes(static_cast<unsigned char>((*s)[token_idx]));
+        assert(token_idx + tlen <= str_len);
+        new (output_data + output_index) std::string(s->substr(token_idx, tlen));
+        ++output_index;
+        token_idx += tlen;
+        ++tokens;
+      }
+      // Padding strings
+      assert(tokens + (mark_ * 2) <= max_tokens);
+      const size_t pads = max_tokens - (mark_ * 2) - tokens;
+      for (size_t p = 0; p < pads; ++p) {
+        new (output_data + output_index) std::string(padvalue_);
+        ++output_index;
+      }
+      if (mark_) {
+        new (output_data + output_index) std::string(&end_text, 1);
+        ++output_index;
+      }
+    }
+  }
+  return Status::OK();
+}
+
+Status onnxruntime::contrib::Tokenizer::SeparatorTokenize(OpKernelContext* ctx,
+                                                          size_t N, size_t C,
+                                                          const std::vector<int64_t>& input_dims) const {
+  using namespace tokenizer_details;
+
+  // We store the length of the original pattern within the
+  // Ternary Tree so on search miss we drop this many characters from the
+  // string
+  struct ValueType {
+    size_t w_len;
+    int priority_;
+  };
+
+  using SearchTree = TernarySearchTree<wchar_t, ValueType>;
+
+  SearchTree stree;
+  std::wstring_convert<std::codecvt_utf8<wchar_t>> converter(conv_error, wconv_error);
+  int priority = 0;  // earlier search patterns get priority
+  for (auto& sep : separators_) {
+    std::wstring wsep = converter.from_bytes(sep);
+    if (wsep.empty()) {
+      return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
+                    "No empty separators allowed");
+    }
+    if (wsep == wconv_error) {
+      return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
+                    "Separator strings contains invalid utf8 chars: " + sep);
+    }
+    stree.put(wsep.c_str(), wsep.length(), {wsep.length(), priority});
+    ++priority;
+  }
+
+  // Scan all strings and attempt to find separators in them
+  // collect all the output tokens here
+  size_t max_tokens = 0;
+  std::vector<std::vector<std::wstring>> tokenized_strings;
+  tokenized_strings.reserve(N * C);
+  auto X = ctx->Input<Tensor>(0);
+  auto const input_data = X->template Data<std::string>();
+  for (size_t n = 0; n < N; ++n) {
+    for (size_t c = 0; c < C; ++c) {
+      const size_t input_idx = (n * C) + c;
+      const auto* s = input_data + input_idx;
+
+      std::wstring wstr = converter.from_bytes(*s);
```

Comment:
I am converting the whole string. It just many of them. If the question is if it is possible to convert many strings in one call, the answer is no.

---
In reply to: [238261318](https://github.com/Microsoft/onnxruntime/pull/31#discussion_r238261318) [](ancestors = 238261318)

Reply:
I cannot access the commit anymore, maybe because it went public yesterday. I think the line std::wstring wstr = converter.from_bytes(*s); could be moved in the loop below on the whole string. As I understood your code, you convert into bytes the Following strings: (input_data + (n*C) + 1), (input_data + + (n*C) + 2), … maybe you could convert input_data + (n*C) just once.

Reply:
I am afraid it is not possible unless you demonstrate what you have in mind. There are separate std::string objects stored in each position. The all have their buffers and most of them are allocated on the heap out of line with Tensor memory. I do not have a continuous buffer to perform the conversion at once.. Nor do I believe it is beneficial the way I understand the spec. Conversion is cheap does no matter I perform it by chunks or not. Allocating new buffers is expensive but I am not seeing the way to avoid it. It also does add inconvenience of tracking the boundaries of separate strings.

---
In reply to: [238688900](https://github.com/Microsoft/onnxruntime/pull/31#discussion_r238688900) [](ancestors = 238688900)